mesonpy/__init__.py

```diff
@@ -542,6 +553,60 @@ def build(self, directory: Path) -> pathlib.Path:
 
         return wheel_file
 
+    def build_editable(self, directory: Path) -> pathlib.Path:
+        self._project.build()  # ensure project is built
+
+        wheel_file = pathlib.Path(directory, f'{self.name}.whl')
+
+        project_path = self._source_dir / '.mesonpy' / 'editable' / 'install'
+        import_paths = {
+            os.fspath(project_path.joinpath(
+                *self._sysconfig_paths[path].split(os.path.sep)
+            ))
+            for path in ('purelib', 'platlib')
+        }
+        top_level_modules = self._project.top_level_modules
+        rebuild_commands = [list(cmd) for cmd in self._project.build_commands(project_path)]
+
+        project_path.mkdir(parents=True, exist_ok=True)
+
+        with mesonpy._wheelfile.WheelFile(wheel_file, 'w') as whl:
+            self._wheel_write_metadata(whl)
+            whl.writestr(
+                f'{self.distinfo_dir}/direct_url.json.py',
+                f'file://{os.fspath(self._project.source_dir)}'.encode(),
+            )
+
+            # install hook module
+            hook_module_name = f'_mesonpy_hook_{self.normalized_name.replace(".", "_")}'
+            hook_install_code = textwrap.dedent(f'''
+                MesonpyFinder.install(
+                    project_path='{project_path}',
+                    import_paths={list(import_paths)},
+                    top_level_modules={list(top_level_modules)},
+                    rebuild_commands={rebuild_commands},
+                )
+            ''').strip().encode()
+            whl.writestr(
+                f'{hook_module_name}.py',
+                (importlib_resources.files('mesonpy') / '_editable.py').read_bytes() + hook_install_code,
```

Comment:
This seems like a perfectly reasonable use of `importlib.resources.read_binary()`, which works on all versions of python without adding more dependencies. I would simply add a `warnings.filterwarnings()` to disable all deprecation notices from `importlib.resources`.

Reply:
Yes, it's reasonable, but I'd rather make the code future-compatible (using a backport for older versions), than the other way round.

Reply:
But it's not making the code future-compatible.

It's adding a dependency in order to avoid a function that is deprecated with no plans to remove. The function that it's avoiding simply makes use of `(files(f), p).read_bytes()` on newer versions of python, so there's absolutely no reason to deprecate it in the first place, despite which, `importlib.resources` prefers to tell people that it's "literally impossible" to use the ***stdlib*** without either version checks or deprecation warnings.

I think version checks are ugly, and so is using a backports package when the lowest supported python has an identically functioning and also more readable API that does everything the project needs. The `files()` API isn't even a good pathlib idiom anyway -- you cannot pass multiple path components to the initializer function. :rofl:

Reply:
Having a simple definition doesn't mean `read_binary()` won't be removed. Using deprecated API is setting yourself up for pain later. My suggestion is always to use the latest API and consider version checks and backports zero cost, they will not even need to be installed on newer versions of Python. Tools like pyupgrade can remove the version checked code for you when you drop a version.

Reply:
Also, I would never filter deprecation warnings to use an old API in new code. If it's removed, you would have been warned, it would be completely self inflicted.

Reply:
No, I mean that the author said it wasn't going to be removed, and also in Jaraco's discuss.python.org thread where the fact that "people dislike this deprecation and think I was unreasonable to add it" was discussed, no one really seemed convinced that it needed to be removed, multiple PyPA or CPython developers chipped in to say they thought it was unfortunate or in one case "So may I know when this slowing down will happen? Because I’m not going to use that module and will actively oppose people using it until then."

I would actually consider it outright reckless to fully and entirely remove the function from the stdlib for at least until python 3.10 is the oldest version of python still supported. Which is only barely not the most recent version of python still supported!

Because that (`pyupgrade --py310-plus`) is exactly when pyupgrade can start migrating that kind of code automatically, even without version checks.

Also -- if you're going to version gate the code anyway, why not simply use the stdlib only, and at the call site, do a version check and use `(files(p), f).read_binary()` on python 3.10, and `read_binary(f, p)` on lower versions. Fewer dependencies == win, and the version check can still be trivially removed by pyupgrade.

Reply:
> Also, I would never filter deprecation warnings to use an old API in new code. If it's removed, you would have been warned, it would be completely self inflicted.
> 
> (The one special case is if you can talk the library into keeping the API and removing the warning in future versions, then you can hide it on older versions)

I would *usually* not filter deprecation warnings to use an old API in new code. This is an exceptional case because the deprecation itself seems unreasonable and a *lot* of people are complaining that the module as a whole is moving "recklessly" fast and making it unreasonably difficult to write acceptable code without simply pretending that it was not added to the stdlib at all until 3.10, which is a really lousy feeling to have to feel about the stdlib.

Reply:
I like `files()` too, especially since it allows me to lookup arbitrary files without adding `__init__.py` to every data directory just so the older API works. :D However, I often cannot use it... (I cannot use it in Meson, so instead I have a pending branch that will go add use of `path()`).

Reply:
version checks are okay, they're a very minor technical debt. taking on another runtime dependency should be avoided, that's a big cost.

Reply:
I'm with @eli-schwartz and @rgommers here. We should strive for minimal dependencies. If we can make this work with an API available on older Python versions, I would go that way.